### PR TITLE
update(RayGun): Send files from both sources in same message

### LINKED ISSUE #1172: feat(Raygun): Send files from local disk and constellation on same message
### 1. To send files from local disk and constellation on same message, is necessary to change code in Warp. So this is the ticket on Warp: https://github.com/Satellite-im/Warp/issues/310?

### PR DESCRIPTION
**Pull request by lgmarchi**
### 1. Now is possible to select files from Storage and from Local Disk and send them on same message

https://github.com/Satellite-im/Uplink/assets/63157656/421f9cb1-2ea7-4209-80de-bc5e3db52787

### Which issue(s) this PR fixes 🔨

- Resolve #1172 

### Additional comments 🎤
This PR need to await merge from [#1168](https://github.com/Satellite-im/Uplink/pull/1168) and #1175

### CODE REVIEW

**Comment by dariusc93** [comment]
Blocked until #1175 is merged

**Comment by lgmarchi** [comment]
Blocked until [#1190](https://github.com/Satellite-im/Uplink/pull/1190) is merged

File: ui/src/layouts/storage/send_files_components.rs
Comment (dariusc93) [inline ui/src/layouts/storage/send_files_components.rs]: You could optimize this by removing the item using `map(|index| f.files_selected_to_send(index))` at the end of position.

File: ui/src/layouts/storage/send_files_components.rs
Comment (dariusc93) [inline ui/src/layouts/storage/send_files_components.rs]: It might be better to do the following

```rust
match location {
    Location::Constellation { path } => path.eq(&file_path),
    _ => false
}
```

File: ui/src/components/chat/compose/chatbar.rs
Comment (dariusc93) [inline ui/src/components/chat/compose/chatbar.rs]: ditto

File: ui/src/components/chat/compose/chatbar.rs
Comment (dariusc93) [inline ui/src/components/chat/compose/chatbar.rs]: Here you could have `filename` return a PathBuf then get the filename into a string in a single string. Would reduce the duplication of code here.

File: ui/src/components/chat/compose/chatbar.rs
Comment (dariusc93) [inline ui/src/components/chat/compose/chatbar.rs]: ditto

File: common/src/state/pending_message.rs
Comment (dariusc93) [inline common/src/state/pending_message.rs]: ditto

File: common/src/warp_runner/manager/commands/constellation_commands.rs
Comment (dariusc93) [inline common/src/warp_runner/manager/commands/constellation_commands.rs:302]: Do not worry about it being in this PR, but the `Directory` and `File` now has a `path` function which you could probably use to track to track the items

File: common/src/state/pending_message.rs
Comment (dariusc93) [inline common/src/state/pending_message.rs]: Maybe this should be a filter_map so we can filter and do something like

```rust
let path = match p {
   Location::Disk { path } => path,
   Location::Constellation { path } => PathBuf::from(path),
};

path.file_name().and_then(OsStr::to_str).map(str::to_string)
```

This would simplify the code and if `PathBuf::file_name` returned `None` it would be filtered out.

File: common/src/warp_runner/manager/commands/constellation_commands.rs
Comment (lgmarchi) [reply common/src/warp_runner/manager/commands/constellation_commands.rs:302]: I think this is different @dariusc93, I use it to navigate opening a new folder, and backing to previous one!

File: ui/src/layouts/storage/send_files_components.rs
Comment (lgmarchi) [reply ui/src/layouts/storage/send_files_components.rs]: I didn't understand what is to do here

File: ui/src/layouts/storage/send_files_components.rs
Comment (lgmarchi) [reply ui/src/layouts/storage/send_files_components.rs]: Done!

File: ui/src/layouts/storage/send_files_components.rs
Comment (dariusc93) [reply ui/src/layouts/storage/send_files_components.rs]: Looking at the code again, it can remain as is.

File: common/src/warp_runner/manager/commands/constellation_commands.rs
Comment (sdwoodbury) [inline common/src/warp_runner/manager/commands/constellation_commands.rs:303]: @lgmarchi do you know where the directory named "un-named directory" comes from? I added a logging statement where `dirs_opened.remove(0)` was called and that's the name of the directory getting removed.

File: ui/src/components/chat/compose/chatbar.rs
Comment (sdwoodbury) [inline ui/src/components/chat/compose/chatbar.rs]: not a blocker but if you `map` `new_files` to a `Location::Disk` first, using `filter` on the files retrieved from `state` results in a much smaller closure. The naming is also better. Also using `drain(..)` removes the need to `clone` the files being iterated over. 

```
if let Some(new_files) = FileDialog::new()
    .set_directory(dirs::home_dir().unwrap_or_default())
    .pick_files()
{
    let new_files: Vec<Location> = new_files.iter()
    .map(|path| Location::Disk { path: path.clone() })
    .collect();

    let mut current_files: Vec<_> =  state.read().get_active_chat().map(|f| f.files_attached_to_send)
    .unwrap_or_default().drain(..).filter(|x| !new_files.contains(x)).collect();
    current_files.extend(new_files);
    state.write().mutate(Action::SetChatAttachments(chat_id, current_files));
    update_send();
}
```

File: ui/src/layouts/storage/mod.rs
Comment (sdwoodbury) [inline ui/src/layouts/storage/mod.rs:82]: I don't like the name `storage_files_to_chat_mode_is_active`. It looks like this variable is used to denote a file picker mode. Something like `is_selecting_files` would be better. 

I don't like the name `on_files_selected_to_send`. Prefer `on_files_attached`. 

After looking at the rest of the file, it looks like a separate Layout is warranted. The FilesLayout is different from a file picker. Combining those two into one view causes clutter and coupling. This seems especially unnecessary in the storage view because the refactor was supposed to separate the control logic from the view.

File: ui/src/layouts/storage/send_files_components.rs
Comment (sdwoodbury) [inline ui/src/layouts/storage/send_files_components.rs]: I think `select_files_to_send_mode` would be better named as `is_selecting_files`.

File: ui/src/layouts/storage/send_files_components.rs
Comment (sdwoodbury) [inline ui/src/layouts/storage/send_files_components.rs]: `add_remove` is the same as `toggle`. `toggle_selected_file` is a better name. 

Also, you should use references for the function arguments if you can. `storage_controller` can be a `&mut` and `file_path` can be a `&str`.

File: ui/src/layouts/storage/send_files_components.rs
Comment (sdwoodbury) [inline ui/src/layouts/storage/send_files_components.rs]: `on_press_send_files_button` could be `on_send`. The name is more concise.

File: common/src/warp_runner/manager/commands/constellation_commands.rs
Comment (lgmarchi) [reply common/src/warp_runner/manager/commands/constellation_commands.rs:303]: I don't know exactly, Darius made a change in Warp to not happen this problem anymore, and I did this code to guarantee it will not happen on front end. 

Anyway, when it happens, this directory is the root directory.

**Comment by luisecm** [comment]
Looks like there is an issue when compiling on MacOS and Windows. I saw from comments above that this ticket is blocked so probably this is expected to happen at this moment:

image

**Comment by dariusc93** [comment]
> Looks like there is an issue when compiling on MacOS and Windows. I saw from comments above that this ticket is blocked so probably this is expected to happen at this moment:
> 
> image

Thats not OS related but was related to a it being mapped to a `Option<_>` instead of filtered using `filter_map` as well as due to the reference from `Location::Disk` not being cloned. I pushed a quick fix for that

**Review by stavares843 — CHANGES_REQUESTED** [review]
Captura de ecrã 2023-09-13, às 22 21 04

clicking on home and go to files button doesn't do anything 

macOS, m1

**Review by stavares843 — COMMENTED** [review]
Captura de ecrã 2023-09-13, às 22 22 21

when you have no files in files appears like this maybe we could add a message saying ''no files available'' or something

**Review by stavares843 — CHANGES_REQUESTED** [review]
if you upload 1 file from the files


Captura de ecrã 2023-09-13, às 22 24 43

and then drag and drop a file from your computer, he will remove the file that was already there

if you do the same flow but instead of dragging and dropping you go to add file > select file from computer > he will add the file instead of removing the previous one - the correct behavior

**Review by stavares843 — CHANGES_REQUESTED** [review]
if you are uploading files into files section 

Captura de ecrã 2023-09-13, às 22 26 37


and go to chat and try to upload a file from the files, the files are all non clickable

Captura de ecrã 2023-09-13, às 22 26 30

**Comment by stavares843** [comment]
all the above was tested in macOS, m1

**Comment by lgmarchi** [comment]
@stavares843 Nice points Sara, About drag and drop I know how to solve, the other one I will see with @XileHorizon best solution for the user.

**Comment by lgmarchi** [comment]
> if you are uploading files into files section
> 
> Captura de ecrã 2023-09-13, às 22 26 37
> and go to chat and try to upload a file from the files, the files are all non clickable
> 
> Captura de ecrã 2023-09-13, às 22 26 30

Fixed @stavares843. 

https://github.com/Satellite-im/Uplink/assets/63157656/3352c542-044c-4ea7-8906-f02c7c7e3597

But I will wait until #1206  is merged, to merge this one.

**Comment by stavares843** [comment]
linter is failing 🔨

**Review by InfamousVague — APPROVED** [review]

**Comment by stavares843** [comment]
Captura de ecrã 2023-09-13, às 22 22 21

when you have no files in files appears like this maybe we could add a message saying ''no files available'' or something

Captura de ecrã 2023-09-13, às 22 21 04

clicking on home and go to files button doesn't do anything 

macOS, m1

these still happens with the latest

**Comment by stavares843** [comment]
but not against fixing them in other PR

**Comment by stavares843** [comment]
linter is failing 🔨

**Comment by lgmarchi** [comment]
> Captura de ecrã 2023-09-13, às 22 22 21
> when you have no files in files appears like this maybe we could add a message saying ''no files available'' or something
> 
> Captura de ecrã 2023-09-13, às 22 21 04
> clicking on home and go to files button doesn't do anything
> 
> macOS, m1
> 
> these still happens with the latest

@stavares843 I understood first problem. This I think we can for a next PR. 

But second one I think it is a bug but I didn't understand, can you explain me better?

**Review by sdwoodbury — COMMENTED** [review]
If the user is picking a file from Constellation, they don't need the following features from `FilesLayout`
- drag/drop
- upload
- rename

It basically just needs folder navigation and the new `send_files_from_chat_topbar` element. 

I hope the folder navigation part can be made into its own element which could then be shared between two different views. 

The FilesLayout is already very large. The `rsx!` macro is 400 lines. I think that making it act as a different view based on a parameter further complicates something which is already difficult to maintain.

**Comment by stavares843** [comment]
lucas and i spoke in discord 🔨

**Comment by InfamousVague** [comment]
> > Captura de ecrã 2023-09-13, às 22 22 21
> > when you have no files in files appears like this maybe we could add a message saying ''no files available'' or something
> > Captura de ecrã 2023-09-13, às 22 21 04
> > clicking on home and go to files button doesn't do anything
> > macOS, m1
> > these still happens with the latest
> 
> @stavares843 I understood first problem. This I think we can for a next PR.
> 
> But second one I think it is a bug but I didn't understand, can you explain me better?

I think she means to say that shes already at home so when clicking home it doesn't do anything, maybe we should disable the home button when we're already at the home / root level?

**Comment by lgmarchi** [comment]
> > > Captura de ecrã 2023-09-13, às 22 22 21
> > > when you have no files in files appears like this maybe we could add a message saying ''no files available'' or something
> > > Captura de ecrã 2023-09-13, às 22 21 04
> > > clicking on home and go to files button doesn't do anything
> > > macOS, m1
> > > these still happens with the latest
> > 
> > 
> > @stavares843 I understood first problem. This I think we can for a next PR.
> > But second one I think it is a bug but I didn't understand, can you explain me better?
> 
> I think she means to say that shes already at home so when clicking home it doesn't do anything, maybe we should disable the home button when we're already at the home / root level?

I got it later, it can be a solution, I would like to hear @XileHorizon about it as well.

@InfamousVague I was thinking in not deactivate it, but if user clicks on it, return a feedback saying that user is already on that folder. Just a suggestion.

**Comment by lgmarchi** [comment]
> If the user is picking a file from Constellation, they don't need the following features from `FilesLayout`
> 
> * drag/drop
> * upload
> * rename
> 
> It basically just needs folder navigation and the new `send_files_from_chat_topbar` element.
> 
> I hope the folder navigation part can be made into its own element which could then be shared between two different views.
> 
> The FilesLayout is already very large. The `rsx!` macro is 400 lines. I think that making it act as a different view based on a parameter further complicates something which is already difficult to maintain.

I will think in a way to separate those layers in a next PR.

**Review by dariusc93 — APPROVED** [review]

**Review by stavares843 — APPROVED** [review]

**Pull request merged**
